Fixes #22285 - Raise error when strong params filters out params

tbrisker opened the pull request:
Currently we silently filter out parameters (or log in production),
causing unexpected results when passing an invalid param (either
incorrect name or incorrect type). This can lead to unexpected results,
since the user, seeing no error, assumes the request was successful when
in fact some of the parameters were filtered out.

tbrisker committed:
Fixes #22285 - Raise error when strong params filters out params

Currently we silently filter out parameters (or log in production),
causing unexpected results when passing an invalid param (either
incorrect name or incorrect type). This can lead to unexpected results,
since the user, seeing no error, assumes the request was successful when
in fact some of the parameters were filtered out.

tbrisker commented:
This currently Breaks All Things. setting to WoC to figure out.

tbrisker commented:
Previously had 636 failing tests, still WIP to fix all broken tests.

tbrisker commented on test/controllers/api/v2/hosts_controller_test.rb:
```diff
@@ -314,7 +314,7 @@ def expect_attribute_modifier(modifier_class, args)
 
   test "should update hostgroup_id of host" do
     @host = FactoryBot.create(:host, basic_attrs_with_hg)
-    put :update, params: { :id => @host.to_param, :hostgroup_id => Hostgroup.last.id }
```
The last hostgroup doesn't necessarily have an environment, which is required, and reset if the hg changes.

iNecas commented:
Could we wait with this change until 1.17 is out + a week or two: many plugin authors haven't had green tests for a while, so would be good to give them some time to actually do other things than catching up with upstream breakages?

tbrisker committed:
Refs #22285 - Fix tests for strong params enforcement

- fix users controller tests
  A user cannot change their own auth_source. Passing illegal
  parameters should raise ActionController::UnpermittedParameters
- Fix auth_source_ldap controller tests

tbrisker commented:
Down to 237 :)

tbrisker commented:
@iNecas I expect fixing all tests and reviewing it will take several days at best. I don't mind waiting a bit but it will be painful to keep up to date over a long period. Besides the impact on users and developers I outlined on [discourse](https://community.theforeman.org/t/strong-parameters-is-borked/8070), this also exposes multiple tests which we thought were passing when in fact they were failing silently - I am still digging into it, but there may be other bugs discovered due to these tests. In other words - the sooner we fix this properly the better. 
That being said, I think for most plugins fixing this would be a matter of adding a few parameters to the filters already in place; any broken tests discovered due to this are actually broken already without us knowing it. Once I finish doing the core part of it, I'll try looking into some of the plugins.

thomasmckay commented:
@waldenraines FYI - This will require significant effort to fix and test the katello UI since when an item is edited its json is often sent back to server with extra fields. Product edit, for example, gets "An error occurred saving the Product: found unpermitted parameters: :cp_id, :created_at, :updated_at". This will be common on all of the pages, I think.

mmoll commented:
@tbrisker please rebase :)

tbrisker commented:
rebased, now lets see how many new test have broken in the last 2 months...

mmoll commented:
@tbrisker this needs a rebase :)

I think this should go in soon, especially now that 1.18 is branched.

mmoll commented:
@tbrisker ping?

tbrisker commented:
@mmoll last time i tried to get this mergeable i figured out there's some refactoring to the host[group] form that's needed to fix one of the issues, GH-5458 is WIP to fix that. Once I finish with that this one is next in line.

ares commented:
:cry:
